src/drug_discovery/structures/protein.py

```diff
@@ -730,6 +634,51 @@ def extract_metals_and_cofactors(self) -> Tuple[list[str], list[str]]:
 
         return metal_resnames, cofactor_resnames
 
+    def extract_ligand(self, exclude_resnames: Optional[set] = None):
+        """
+        Extracts ligand(s) from a PDB file using RDKit.
+
+        Args:
+            exclude_resnames (set): Residue names to exclude (e.g., water).
+
+        """
+
+        from rdkit import Chem
+
+        if exclude_resnames is None:
+            exclude_resnames = {"HOH"}
+
+        ligand_lines = []
+        with open(self.file_path, "r") as f:
+            for line in f:
+                if line.startswith("HETATM"):
+                    resname = line[17:20].strip()
+                    if resname not in exclude_resnames:
+                        ligand_lines.append(line)
+
+        if not ligand_lines:
+            raise ValueError("No ligand HETATM records found in the PDB.")
+
+        # Write ligand atoms to a temporary file
+        tmp = tempfile.NamedTemporaryFile(mode="w+", suffix=".pdb", delete=False)
+        try:
+            for line in ligand_lines:
+                tmp.write(line)
+            tmp.write("END\n")
+            tmp.flush()
+            tmp.seek(0)
+            ligand_pdb_block = tmp.read()
+        finally:
+            tmp.close()
+            os.unlink(tmp.name)
+
+        # Parse with RDKit
+        mol = Chem.MolFromPDBBlock(ligand_pdb_block, sanitize=True, removeHs=False)
```

Comment:
Don’t round‑trip through a temp file. You already call MolFromPDBBlock; just build a PDB block in memory.